### DISCUSSION TITLE
Unreleased references.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#294] [Q&A] [answered]
I've removed any rendering code and so I'm just looking at my core graphics shutdown.

Heres' the shutdown code (I think it's based on what's in the samples)..
```
        protected void OnDispose()
        {
            Debug.LogMessage("Device3d.OnDispose");
            DxImmediateContext.ClearState();
            DxImmediateContext.Flush();
            DxImmediateContext.Release();
            DxImmediateContext.Dispose();
            DxImmediateContext = null;

#if DEBUG
            uint refCount = DxDevice.Release();
            if (refCount > 0)
            {
                System.Diagnostics.Debug.WriteLine($"Direct3D11: There are {refCount} unreleased references left on the device");

                ID3D11Debug? d3d11Debug = DxDevice.QueryInterfaceOrNull<ID3D11Debug>();
                if (d3d11Debug != null)
                {
                    d3d11Debug.ReportLiveDeviceObjects(ReportLiveDeviceObjectFlags.Detail | ReportLiveDeviceObjectFlags.IgnoreInternal);
                    d3d11Debug.Dispose();
                }
            }
#else
        DxDevice.Dispose();
#endif
            DxAdapter.Dispose();
        }
```
        
If I just run the app and shut it down with the window-close-button I get the following output..

> T: 1 | MainWindow.MainWindow_Closed
> T: 4 | Engine | MainLoop thread exiting
> T: 4 | Swapchain.OnDispose
> T: 4 | Device3d.OnDispose
> **_Direct3D11: There are 1 unreleased references left on the device_**
> The program '[29020] DemoCube.exe' has exited with code 0 (0x0).

But if I set a breakpoint on the line, giving the app a moments pause and then allow the app to continue and complete.
`uint refCount = DxDevice.Release();`

I get the following output..

> T: 1 | MainWindow.MainWindow_Closed
> T: 4 | Engine | MainLoop thread exiting
> T: 4 | Swapchain.OnDispose
> T: 4 | Device3d.OnDispose
> The program '[28928] DemoCube.exe' has exited with code 0 (0x0).

It seems like some reference is taking some time to release. Notice that I'm not getting the usual live objects output, so what does this mean? And should I care? I don't like seeing it there, because usually if I don't address these things early it just becomes more difficult later to track down unreleased resources.

**Comment by amerkoleci** [comment, marked as answer]
Strange, you can take look at example here:
https://github.com/amerkoleci/Vortice.Windows/tree/main/src/samples/HelloDirect3D11

I have 0 unreleased referenced, I would suggest to copy code from there, make sure to enable native debugging as well.

**Reply by Gavin-Williams** [reply]
I'll mark this is an answer, I think I have it working a bit better now. Someone on Discord helped as well with how to use the InfoQueue to filter out some info.
